temporal-testing/src/main/java/io/temporal/internal/testservice/TestWorkflowMutableStateImpl.java

```diff
@@ -2115,6 +2126,209 @@ public void completeQuery(QueryId queryId, RespondQueryTaskCompletedRequest comp
+     * in pendingActivity.activityTask. The ActivityTaskData itself also has some of what we need.
+     */
+    ActivityTaskData pendingActivity = sm.getData();
+    PollActivityTaskQueueResponse.Builder task = pendingActivity.activityTask.getTask();
+    ActivityTaskScheduledEventAttributes scheduledEvent = pendingActivity.scheduledEvent;
+
+    State state = sm.getState();
+    PendingActivityInfo.Builder builder =
+        PendingActivityInfo.newBuilder()
+            .setActivityId(scheduledEvent.getActivityId())
+            .setActivityType(scheduledEvent.getActivityType())
+            .setState(computeActivityState(state, pendingActivity));
+
+    // In golang, we set one but never both of these fields, depending on the activity state
+    if (builder.getState() == PendingActivityState.PENDING_ACTIVITY_STATE_SCHEDULED) {
+      builder.setScheduledTime(task.getScheduledTime());
+    } else {
+      builder.setLastStartedTime(task.getStartedTime());
+      builder.setLastHeartbeatTime(task.getStartedTime());
+    }
+
+    if (pendingActivity.lastHeartbeatTime > 0) {
+      // This may overwrite the heartbeat time we just set - that's fine
+      builder.setLastHeartbeatTime(timestampFromMillis(pendingActivity.lastHeartbeatTime));
+
+      if (pendingActivity.heartbeatDetails != null) {
+        builder.setHeartbeatDetails(pendingActivity.heartbeatDetails);
+      }
+    }
+
+    if (scheduledEvent.hasRetryPolicy()) {
+      builder.setAttempt(pendingActivity.retryState.getAttempt());
+      builder.setExpirationTime(pendingActivity.retryState.getExpirationTime());
+      builder.setMaximumAttempts(scheduledEvent.getRetryPolicy().getMaximumAttempts());
+      pendingActivity.retryState.getPreviousRunFailure().ifPresent(builder::setLastFailure);
+      // We don't track this in the test environment right now, but we could.
+      builder.setLastWorkerIdentity("test-environment-worker-identity");
+    } else {
+      builder.setAttempt(1);
+    }
+
+    return builder.build();
+  }
+
+  private Timestamp timestampFromMillis(long millis) {
+    return Timestamp.newBuilder()
+        .setSeconds(millis / 1000)
+        .setNanos(Math.toIntExact(Duration.ofMillis(millis % 1000).toNanos()))
+        .build();
+  }
+
+  // Mimics golang in HistoryEngine.DescribeWorkflowExecution. Note that this only covers pending
+  // states, so there's quite a bit of state-space that doesn't need to be mapped.
+  private PendingActivityState computeActivityState(State state, ActivityTaskData pendingActivity) {
+    if (state == State.CANCELLATION_REQUESTED) {
+      return PendingActivityState.PENDING_ACTIVITY_STATE_CANCEL_REQUESTED;
+    } else if (pendingActivity.startedEvent != null) {
+      return PendingActivityState.PENDING_ACTIVITY_STATE_STARTED;
+    } else {
+      return PendingActivityState.PENDING_ACTIVITY_STATE_SCHEDULED;
+    }
+  }
+
+  private void setTimestampsBasedOnHistory(
```

Comment:
This looks like a setter. Please consider renaming to something like `fillExecutionInfoTimestampsBasedOnHistory` and making this method, `getStartEvent` and `getCompletionEvent` static.